scalafix-reflect/src/main/scala-3/scalafix/internal/reflect/RuleCompiler.scala

```diff
@@ -0,0 +1,71 @@
+package scalafix.internal.reflect
+
+import scalafix.internal.reflect.exception.CompilationException
+
+import java.io.File
+import java.nio.file.Paths
+import scala.meta.io.AbsolutePath
+
+import dotty.tools.dotc.Compiler
+import dotty.tools.dotc.Run
+import dotty.tools.dotc.util.SourceFile
+import dotty.tools.dotc.core.Contexts.NoContext.given_Context
+import dotty.tools.dotc.reporting.StoreReporter
+
+import scala.reflect.io.VirtualDirectory
+import scala.reflect.io.AbstractFile
+import scala.reflect.internal.util.AbstractFileClassLoader
+
+import metaconfig.Configured
+import metaconfig.Input
+
+class RuleCompiler(
+    classpath: String,
+    target: AbstractFile = new VirtualDirectory("(memory)", None)
+) {
+  // we are not configuring any settings!
+  // private val settings = new Settings()
+  // settings.deprecation.value = true // enable detailed deprecation warnings
+  // settings.unchecked.value = true // enable detailed unchecked warnings
+  // settings.outputDirs.setSingleOutput(target)
+  // settings.classpath.value = classpath
+
+  private lazy val reporter: StoreReporter = new StoreReporter()
+  private val compiler: Compiler = new Compiler()
+  private val classLoader: AbstractFileClassLoader =
+    new AbstractFileClassLoader(target, this.getClass.getClassLoader)
+
+  @throws(classOf[CompilationException])
+  def compile(input: Input): Configured[ClassLoader] = {
+    val run: Run = compiler.newRun
+
+    run.compileSources(
+      List(
+        new SourceFile(
+          dotty.tools.io.AbstractFile.getFile(input.path),
+          input.chars
+        )
+      )
+    )
+
+    if (reporter.hasErrors) {
+      val defaultMsg = "Multiple compilation errors occurred for Scala 3, " +
+        "please try to use a CLI version which compiles into Scala 2"
```

Comment:
This would be quite abstract/hard to understand for an end-user.

What about something like:
> Error compiling rule(s) from source using Scala 3 compiler; to use the Scala 2.x compiler instead, use the corresponding scalafix-cli artifact or force scalafixScalaBinaryVersion to 2.x in your build tool.

It's not perfect as scalafixScalaBinaryVersion is a sbt-scalafix concept, but
1. judging by the number of tickets/requests, sbt-scalafix is by far the most used client so far
1. build tools will have something similar (although maybe not named exactly the same) if they allow picking up a specific scalafix scala versiom.

Reply:
I don't have a strong opinion on this tbh, I don't see any problem with requiring users to set a `scalafixScalaBinaryVersion` as well. Could be useful to document this config somewhere tho.